File: src/abstract/Command.ts
import {Channel, invoke} from "@tauri-apps/api/core";
import {EventEmitter} from "./EventEmitter"

export interface TerminatedPayload {
    code: number | null
    signal: number | null
}

type CommandEvent =
    | {
    event: "stdout",
    payload: string
}
    | {
    event: "error",
    payload: string
}
    | {
    event: "terminated",
    payload: TerminatedPayload
}

interface CommandEvents {
    stdout: string,
    error: string,
    terminated: TerminatedPayload
}

export class Command extends EventEmitter<CommandEvents> {
    private readonly command: string
    private readonly args: string[]
    private readonly hide: boolean

    static create(
        program: string,
        args: string | string[] = [],
        hide: boolean = false
    ): Command {
        return new Command(program, args, hide)
    }

    constructor(command: string, args: string | string[] = [], hide: boolean) {
        super();
        this.command = command
        this.args = typeof args === 'string' ? [args] : args
        this.hide = hide
    }

    async spawn(): Promise<Child> {
        const command = this.command
        const args = this.args
        const hide = this.hide

        if (typeof args === 'object') {
            Object.freeze(args)
        }

        const onEvent = new Channel<CommandEvent>()
        onEvent.onmessage = (event) => {
            switch (event.event) {
                case "stdout":
                    this.emit("stdout", event.payload)
                    break
                case "error":
                    this.emit("error", event.payload)
                    break
                case "terminated":
                    this.emit("terminated", event.payload)
                    break
            }
        }
        return await invoke<number>("command_spawn", {
            command,
            args,
            hide,
            onEvent
        }).then((pid) => new Child(pid))
    }
}

export class Child {
    pid: number

    constructor(pid: number) {
        this.pid = pid
    }

    async kill(): Promise<void> {
        await invoke("child_kill", {
            pid: this.pid
        })
    }
}

File: src-tauri/src/tools/command.rs
use std::collections::HashMap;
use std::io::{BufRead, BufReader};
use std::os::windows::process::CommandExt;
use std::process::{Child, Command};
use std::sync::{Arc, Mutex};
use tauri::ipc::Channel;
use tauri::State;

#[derive(Clone, serde::Serialize)]
#[serde(rename_all = "camelCase", tag = "event", content = "payload")]
pub(crate) enum CommandEvent {
    #[serde(rename_all = "camelCase")]
    Stdout(String),
    #[serde(rename_all = "camelCase")]
    Error(String),
    #[serde(rename_all = "camelCase")]
    Terminated {
        code: Option<i32>,
        signal: Option<i32>,
    },
}

pub(crate) struct ChildrenManager {
    pub(crate) children: Arc<Mutex<HashMap<u32, Child>>>,
}

#[tauri::command]
pub(crate) fn command_spawn(
    command: String,
    args: Vec<String>,
    hide: bool,
    on_event: Channel<CommandEvent>,
    manager: State<ChildrenManager>,
) -> Result<u32, ()> {
    let mut binding = Command::new(command);
    let command = binding.args(args);
    command.stderr(std::process::Stdio::piped());
    // 隐藏状态下截取标准输出，否则仅截取错误输出
    if hide {
        command.creation_flags(0x08000000);
        command.stdout(std::process::Stdio::piped());
    }
    let mut child = match command.spawn() {
        Ok(child) => child,
        Err(e) => {
            on_event.send(CommandEvent::Error(e.to_string())).unwrap();
            on_event
                .send(CommandEvent::Terminated {
                    code: None,
                    signal: None,
                })
                .unwrap();
            return Err(());
        }
    };
    let pid = child.id();
    let stderr = child.stderr.take().unwrap();
    // 创建一个线程，用于读取子进程的标准输出
    if hide {
        let stdout = child.stdout.take().unwrap();
        let on_event = on_event.clone();
        std::thread::spawn(move || {
            let mut reader = BufReader::new(stdout).lines();
            while let Some(Ok(line)) = reader.next() {
                // 处理每一行输出
                on_event.send(CommandEvent::Stdout(line)).unwrap();
            }
        });
    }
    manager.children.lock().unwrap().insert(pid, child);
    let children = manager.children.clone();
    // 创建一个线程，用于读取子进程的错误输出
    std::thread::spawn(move || {
        let mut reader = BufReader::new(stderr).lines();

        while let Some(Ok(line)) = reader.next() {
            // 处理每一行输出
            on_event.send(CommandEvent::Error(line)).unwrap();
        }

        match children.lock().unwrap().remove(&pid) {
            None => {
                on_event
                    .send(CommandEvent::Terminated {
                        code: None,
                        signal: None,
                    })
                    .unwrap();
            }
            Some(mut child) => {
                let _ = child.kill();
                let code = child.wait().ok();
                on_event
                    .send(CommandEvent::Terminated {
                        code: code.and_then(|status| status.code()),
                        signal: None,
                    })
                    .unwrap();
            }
        }
    });
    Ok(pid)
}

#[tauri::command]
pub(crate) fn child_kill(id: u32, manager: State<ChildrenManager>) {
    match manager.children.lock().unwrap().remove(&id) {
        None => {}
        Some(mut child) => {
            let _ = child.kill();
        }
    }
}
